### ISSUE TITLE
[Feature] Towards solving Jobshop & RCPSP

### ISSUE THREAD

**Issue by Dekker1** [huub-solver/huub#23]
A first target for the solver is to be able to (relatively efficiently) solve (variants of) the jobshop and RCPSP problems. We have discussed the following steps as part of the process of getting there.

- [x] Propagators for remaining constraints
  - [x] Integer element
  - [x] Reified linear
- [x] Search heuristics
  - [x] Warm Start
  - [x] Disable CP search after restart

Further improvements could then likely be achieved by implementing specialized handling for `disjunctive` and `cumulative`

**Comment by AllenZzw** [comment]
For some reason, the following instance trigger some error in the solver: 
```
int: n_jobs = 3;
int: n_machines = 3;
int: n_tasks_per_job = n_machines;      % Each job has one task per machine.
set of int: jobs = 1..n_jobs;
set of int: tasks = 1..n_tasks_per_job;

array [jobs, tasks] of 0..(n_machines-1): job_task_machine = array2d(jobs, tasks, [
    2, 0, 1,
    0, 1, 2,
    0, 2, 1,
]);
array [jobs, tasks] of int: job_task_duration = array2d(jobs, tasks, [
    13, 24, 7,
    8, 13, 14,
    22, 17, 9,
]);

int: min_duration = 
    max([sum([job_task_duration[i, j] | j in tasks]) | i in jobs]);

int: max_duration = sum([job_task_duration[i, j] | i in jobs, j in tasks]);

%-----------------------------------------------------------------------------%
% Model variables.
%

array [jobs, tasks] of var 0.. max_duration: job_task_start;
var min_duration..max_duration: t_end;

%-----------------------------------------------------------------------------%
% Constraints.
%

constraint
    forall ( j in jobs, k in tasks ) (
        job_task_duration[j, k]  >=  0
    );

constraint
    forall ( j in jobs, k in 1..(n_tasks_per_job - 1) ) (
        job_task_start[j, k] + job_task_duration[j, k]  <=
            job_task_start[j, k + 1]
    );

constraint
    forall ( j in jobs ) (
        0  <=  job_task_start[j, 1]
    );

constraint
    forall (
        ja in jobs,
        jb in (ja + 1)..n_jobs,
        ka, kb in tasks
    ) (
        % (N.B.: if-then-elses flatten somewhat faster than implications.)
        if
           job_task_machine[ja, ka]  =  job_task_machine[jb, kb]
        then
            no_overlap( 
                job_task_start[ja, ka], job_task_duration[ja, ka],
                job_task_start[jb, kb], job_task_duration[jb, kb]
            )
        else
            true
        endif
    );

predicate no_overlap(var int: t_a, var int: d_a, var int: t_b, var int: d_b)  =
    ( t_a + d_a  <=  t_b )  \/  ( t_b + d_b  <=  t_a );

constraint
    forall ( j in jobs ) (
            job_task_start[j, n_tasks_per_job] +
            job_task_duration[j, n_tasks_per_job]
        <=
            t_end
    );

solve minimize t_end;
```

**Comment by AllenZzw** [comment]
The issue is about the reason of setting integer upper bounds in the linear propagator. It has been fixed in #35 .
